### PR TITLE
 feat(datepicker): enable user to define his own date structure 

### PR DESCRIPTION
**Pull request by msosa**
Before submitting a pull request, please make sure you have at least performed the following:

 - [x] read and followed the [CONTRIBUTING.md](https://github.com/ng-bootstrap/ng-bootstrap/blob/master/CONTRIBUTING.md) guide.
 - [x] built and tested the changes locally.
 - [x] added/updated any applicable tests.
 - [x] added/updated any applicable API documentation.
 - [x] added/updated any applicable demos.

This is a continuation of #1816 since the original requester(@edriang) said he will no longer work on it. I added the changes requested by @maxokorokov 

Closes #1816

### CODE REVIEW

File: src/datepicker/ngb-date-adapter.ts [hunk 1]
```diff
@@ -0,0 +1,47 @@
+import {Injectable} from '@angular/core';
+import {NgbDateStruct} from './ngb-date-struct';
+
+/**
+ * Abstract type serving as a DI token for the service converting from your application Date model to internal NgbDate
```
Comment (maxokorokov) [inline src/datepicker/ngb-date-adapter.ts]: typo: `to internal NgbDate model` → `to internal NgbDateStruct model`

File: src/datepicker/ngb-date.spec.ts [hunk 2]
```diff
@@ -61,3 +62,42 @@ describe('ngb-date', () => {
     });
   });
 });
+
+describe('ngb-date model adapter', () => {
```
Comment (maxokorokov) [inline src/datepicker/ngb-date.spec.ts]: please move to a separate `ngb-date-adapter.spec.ts` file

File: src/datepicker/ngb-date.spec.ts [hunk 3]
```diff
@@ -61,3 +62,42 @@ describe('ngb-date', () => {
...
+  describe('toModel', () => {
+
+    it('should convert null and undefined as null', () => {
+      expect(adapter.toModel(null)).toBe(null);
+      expect(adapter.toModel(undefined)).toBe(null);
+    });
+
+    it('should convert a valid date', () => {
+      expect(JSON.stringify(adapter.toModel({year: 2016, month: 10, day: 15})))
+          .toBe(JSON.stringify({year: 2016, month: 10, day: 15}));
```
Comment (maxokorokov) [inline src/datepicker/ngb-date.spec.ts]: I think you can use `.toEqual` instead of stringifying

**Comment by msosa** [comment]
updated, was wondering if I should add another test to ngb-date-adapter maybe using the Date class

**Comment by maxokorokov** [comment]
Thanks again, @msosa. I took the liberty of squashing and rebasing for cleaner history. Also there was a small conflict to resolve.

And then I realized you were opening the PR form your master branch :/ Sorry about that

Will merge when Travis is green

**Comment by msosa** [comment]
No worries, thanks!

**Pull request merged**